src/accounts.rs

```diff
@@ -319,17 +328,72 @@ struct InnerConfig {
     pub accounts: Vec<AccountConfig>,
 }
 
+impl Drop for Config {
+    fn drop(&mut self) {
+        if let Some(lock_task) = self.lock_task.take() {
+            lock_task.abort();
+        }
+    }
+}
+
 impl Config {
-    /// Creates a new configuration file in the given account manager directory.
-    pub async fn new(dir: &Path) -> Result<Self> {
+    /// Creates a new Config for `file`, but doesn't open/sync it.
+    async fn new_nosync(file: PathBuf, lock: bool) -> Result<Self> {
+        let dir = file.parent().context("can't get config file directory")?;
         let inner = InnerConfig {
             accounts: Vec::new(),
             selected_account: 0,
             next_id: 1,
         };
-        let file = dir.join(CONFIG_NAME);
-        let mut cfg = Self { file, inner };
+        if !lock {
+            let cfg = Self {
+                file,
+                inner,
+                lock_task: None,
+            };
+            return Ok(cfg);
+        }
+        let lockfile = dir.join(LOCKFILE_NAME);
+        let mut lock = fd_lock::RwLock::new(fs::File::create(lockfile).await?);
+        let (locked_tx, locked_rx) = oneshot::channel();
+        let lock_task: JoinHandle<anyhow::Result<()>> = tokio::spawn(async move {
+            let mut timeout = Duration::from_millis(100);
+            let _guard = loop {
+                match lock.try_write() {
+                    Ok(guard) => break Ok(guard),
+                    Err(err) => {
+                        if timeout.as_millis() > 1600 {
+                            break Err(err);
+                        }
+                        // This adds delays to the tests, but otherwise ok.
+                        sleep(timeout).await;
```

Comment:
But the code that executes drop() is really async. So, i can use blocking_recv() there, it will panic as per https://docs.rs/tokio/latest/tokio/sync/oneshot/struct.Receiver.html#method.blocking_recv :
> This function panics if called within an asynchronous execution context.

So, all the ugly code here with this async locking task and sleep() is because of ugly fd_lock RAII guards that otherwise would make Config a self-referential struct. If u have any ideas how to w/a that, it would be nice.

Reply:
When i tried to use [futures](https://docs.rs/futures/latest/futures/index.html)::[executor](https://docs.rs/futures/latest/futures/executor/index.html)::[block_on](https://docs.rs/futures/latest/futures/executor/fn.block_on.html#), it deadlocked if there's the only tokio worker thread. Even if there would be more threads, deadlocks are possible, but just less probable

Reply:
I think this 100-ms sleep is ok though, we don't open configs for writing often ufaiu. I just don't like that the code is so complicated

Reply:
We need to wait for the [possibly existing] previous instance of this locking task to release the lock. It's async, so i have no idea how to get rid of sleep()